### ISSUE TITLE
melt.data.table inconsistent with reshape2::melt

### ISSUE THREAD

**Issue by MichaelChirico** [Rdatatable/data.table#1699]
See this [SO Q&A](http://stackoverflow.com/q/37172700/3576984).

Here's what works in `reshape2`:

``` r
buyer      <- c("A", "A", "B", "A", "B", "C")
seller     <- c("C", "B", "C", "B", "C", "A")
negotiator <- c("B", "C", "D", "D", "A", "B")

df <- data.frame(buyer, seller, negotiator)
reshape2::melt(df,0)
#      variable value
#1       buyer     A
#2       buyer     A
#3       buyer     B
#4       buyer     A
#5       buyer     B
#6       buyer     C
#7      seller     C
#8      seller     B
#9      seller     C
#10     seller     B
#11     seller     C
#12     seller     A
#13 negotiator     B
#14 negotiator     C
#15 negotiator     D
#16 negotiator     D
#17 negotiator     A
#18 negotiator     B
```

But that's an error in `melt.data.table`:

``` r
setDT(df)
melt(df, 0)
```

> Error in `melt.data.table(df, 0)` : 
>   One or more values in `id.vars` is invalid.

This is fine, as we can simply set `measure.vars = names(df)`, but IIUC we're aiming for `melt.data.table` to subsume `reshape2::melt`.

**Comment by arunsrinivasan** [comment]
Michael, thanks. But I'm quite happy with current behaviour of returning an error. `melt(df, -1)` returns something else. It's hard to keep up with undocumented behaviours.

**Comment by MichaelChirico** [comment]
Sounds good to me. Just figured I'd report. I also noticed `reshape2::melt(df, -1)` doing strange things.

**Comment by MichaelChirico** [comment]
Perhaps worth adding a note to `?melt`??

**Comment by arunsrinivasan** [comment]
Sure. A separate Note section should be fine.

**Comment by MichaelChirico** [comment]
Was putting this off because I first wanted to dive into `reshape2` to figure out what's going on. For posterity, here's the relevant snippet from `reshape2:::melt_check`:

```
varnames <- names(data)
if (!missing(id.vars) && is.numeric(id.vars)) {
    id.vars <- varnames[id.vars]
}
# ...
measure.vars <- setdiff(varnames, id.vars)
```

So `id.vars` becomes `character(0)` when `id.vars = 0`, and in turn `measure.vars` becomes exactly `varnames`. This also helps explain what's happening when `id.vars = -1` -- `id.vars` becomes `varnames[-1]`, i.e., all but the last column.

**Comment by mattdowle** [comment]
Closed by 64374174b7639fd53f18b8097fe48626d3ea3603 in https://github.com/Rdatatable/data.table/pull/2290.
Think the space between # and issue number in the commit message prevented auto close and link.
